lighthouse-core/lib/dom-helpers.js

```diff
@@ -0,0 +1,52 @@
+/**
+ * @license
+ * Copyright 2017 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+
+/**
+ * @fileoverview Helpers to be used with driver.evaluateAsync(). That is, these
+ * methods are intended to be injected into and run in the context of the page.
+ */
+
+/* global document */
+
+/**
+ * @param {string=} selector A simple CSS select to filter nodes on. Combinators
+ * are not supported.
+ * @param {!Array<!Element>}
+ */
+function getElementsInDocument(selector) {
+  const results = [];
+
+  const _findAllElements = nodes => {
+    for (let i = 0, el; el = nodes[i]; ++i) {
+      if (!selector || el.matches(selector)) {
+        results.push(el);
+      }
+      // If the element has a shadow root, dig deeper.
+      if (el.shadowRoot) {
+        _findAllElements(el.shadowRoot.querySelectorAll('*'));
+      }
+    }
+  };
+  _findAllElements(document.querySelectorAll('*'));
+
+  return results;
+}
+
+module.exports = {
+  getElementsInDocument
```

Comment:
since this is only going to be consumed as a string, i think we should toString it at this point. We can append a suffix to help clarify that this is for clientside. perhaps `getElementsInDocumentFn`?